🚀 4단계 - 리팩터링

jivebreaddev opened the pull request:
## 4단계 요구사항 정리

![image](https://github.com/user-attachments/assets/aba93e7b-4169-43a7-9f04-5dfb2515b99a)


## 4단계 요구사항 정리
- [x] 요구사항 1 - HttpRequest 클래스 구현하기 
  - [x] 테스트도 추가하기
- [x] 요구사항 2 - HttpResponse 클래스 구현하기
  - [x] 테스트도 추가하기
- [x] 요구사항 3 - Controller 인터페이스 추가하기
  - [x] Controller Interface 주입하기
- [x] 피드백 사항
  - [x] 사용되지 않는 코드 삭제
  - [x] 프로바이더 객체에서 구현체 정보 입력 -> Config 레이어 생성하기 및 Tomcat에 provider 건네주기 (의존성 분리)
  - [x] 팩토리 명명법 변경하기
  - [x] 블럭 생략 제거
  - [x] query parsing 책임 strategy에 있는 부분 변경하기
  - [x] 상수와 변수 개행
  - [x] Response Header 와 Response 생성로직 추상화 -> HTTP ENTITY
  - [x] queryString 에서 map으로 파싱하는 로직을 리팩토링

## 질문 및 세부구현 정리

- RequestMapping 인터페이스와 Manager 가 Client에 의존하는 문제를 수정하였습니다.
- 확실히 Config레이어를 생성해서 Tomcat 객체로 건네주는 형식으로 깔끔해진것같습니다. (조립과 사용 분리)
- Tomcat과 Servlet 구현의 관계에 대해 이해가 가네요!


정말 재밌었습니다! 
피드백 주시면서 코드의 품질이 올라가는게 느껴지네요!
감사합니다!~

https://github.com/next-step/java-http/pull/73

Jivebread committed:
docs(README): 요구사항 정리

jivebreaddev committed:
feature(Style): 피드백 반영 - 상수와 변수 개행

jivebreaddev committed:
feature(Style): 피드백 반영 - 상수와 변수 개행

jivebreaddev committed:
feature(Style): 피드백 반영 -블럭 생략 지양

jivebreaddev committed:
feature(Style): 피드백 반영 - PERIOD -> DOT으로 명명 변경

jivebreaddev committed:
feature(Config): 피드백 반영 - 의존성 확인 client -> tomcat 의존성 방향 생성 + Config 생성으로 조합과 사용 분리

- @Configuration을 생성해서 조합과 사용 분리하였습니다.
- Tomcat에 Provider를 건네주어서 의존성의 방향을 client -> Tomcat으로 가게끔 변경하였습니다.

jivebreaddev committed:
feature(Style): 피드백 반영 - 명명법 변경하였습니다.
- ControllerLoginFactory -> LoginControllerFactory

jivebreaddev committed:
feature(HttpEntity): 피드백 반영 - HttpEntity로 HttpResponse 생성을 생성 팩토리로 간소화 하였습니다.

jivebreaddev committed:
feature(MapUtil): 피드백 반영 - 중복로직 리팩토링
- MapUtil로 Parsing 로직 이전하였습니다.

jivebreaddev committed:
feature(MapUtil): 피드백 반영 - RequestBody의 경우 parsing의 책임을 parser 로 이전합니다.
-> Parsing 시의 케이스는 추후에 맞게 처리합니다.

jivebreaddev committed:
feature(MapUtil): 요구사항 - RequestMapping 완성하였습니다.
Test들은 이미 추가되어있습니다.

catsbi commented on README.md:
```diff
@@ -113,3 +113,22 @@
   - [x] strategy -> factory provider GET 과 POST에 따라 재구성하기
   - [x] 구성요소에 따른 패키지 분리하기
+
+## 4단계 요구사항 정리
```
늘 readme 깔끔하게 작성해주셔서 저도 다시 한번 체크하기 편하네요 💯

jivebreaddev replied:
늘 칭찬해주셔서 감사합니다!

catsbi commented on tomcat/src/main/java/camp/nextstep/config/ControllerFactoryProviderConfig.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Map;
+
+@Configuration
+public class ControllerFactoryProviderConfig {
```
해당 설정 정보를 호출한 뒤 다음과 같이 호출할 경우 문제가 있을 것 같아요 🤔 

```java
ControllerFactoryProviderConfig config = new ControllerFactoryProviderConfig();

IntStream.closed(0, Integer.MAX_VALUE).mapToObj(idx -> config.ControllerFactoryProviderConfig()).collect(toList());
```

jivebreaddev replied:
caching 을 해두는 방향으로 해야겠네요! 혹은 singleton으로 생성하겠습니다!

catsbi commented on tomcat/src/main/java/camp/nextstep/controller/RequestMapping.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.regex.Pattern;
+import org.apache.coyote.controller.ControllerFactory;
+
+public class RequestMapping implements org.apache.coyote.controller.RequestMapping {
```
이름이 같은 인터페이스와 구현체라니 사용에 문제가 될 수 있습니다. 당장 이 코드만봐도 인터페이스를 지칭하기위해 FQCN으로 작성되었네요 🤔  좀 더 구체적인 이름을 지어도 좋을 것 같네요!

jivebreaddev replied:
ControllerRequestMapping 으로 구현체 이름 변경하였습니다!

catsbi commented on tomcat/src/main/java/camp/nextstep/controller/RequestMapping.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public ControllerFactory getController(String url) {
+        if (STATIC.matcher(url).matches()) {
```
Matcher의 matches()와 find()는 얼핏 동일한 결과를 반환하는 것 같은데 언제 무엇을 써야할지에 대해 확실히 알고 사용하면 좋습니다 👍

jivebreaddev replied:
- matches는 전체의 String에서 regex 패턴을 매칭합니다.
    - 그와 별개로 find는 Substring 에서 Matching 을 확인합니다. -> /sdagagdex.dsifjas/index 
      - 이 경우에 find는 /index와 매칭이 됩니다.

catsbi commented on tomcat/src/main/java/camp/nextstep/controller/DefaultControllerFactory.java:
```diff
@@ -10,26 +10,24 @@
+            .statusLine(ProtocolVersion.HTTP11.getVersion(), StatusCode.OK.name())
+            .responseHeader(responseHeader)
+            .messageBody(message.getBytes())
+            .build();
```
여기서 좀 더 리팩토링하면 다음과 같은 모습도 나올 수 있을 것 같네요. 
```java
return HttpResponse.ok().headers(headers).body(message);
```

콘텐츠의 길이등을 체크해서 세팅해주는건 Response에서 실제로 값을 write하는 시점에 계산해도 충분할 것 같아요. ㅎㅎ

jivebreaddev replied:
HttpResponse.ok().headers(headers).body(message);

이 가능하게끔 변경하였습니다!
축약되니깐 보기 편하네요!

catsbi commented on tomcat/src/main/java/camp/nextstep/controller/RequestMapping.java:
```diff
@@ -0,0 +1,32 @@
+
+    private final Map<String, ControllerFactory> factories;
+    private final ControllerFactory notFoundFactory;
+    private final ControllerFactory resourceFacotry;
```
factories가 있는데 별도의 컨트롤러는 구분해서 필드로 있어야 하는 이유가 있을까요?! 
만약, 예외 목적이라면 notFound밖에는 존재하지 않을까요? 더 확장성을 제공할순 없을까 싶네요!

jivebreaddev replied:
factories에 notFoundFactory 와 resourceFactory를 넣어서 사용하는 방향으로 수정하였습니다.

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/HttpEntity.java:
```diff
@@ -0,0 +1,122 @@
+            .build();
+    }
+
+    public static HttpResponse resource(String version, String url){
```
HttpEntity클래스인데 내부적으로 HttpResponse를 생성해주고 있네요. 객체에 적절한 책임이 맞을까요 ?!

HttpEntity가 좀 더 포괄적인 범위라면 반대로 HttpResponse나 Request에서 HttpEntity를 상속 혹은 위임 혹은 참조하는게 맞지 않을까 싶네요!

jivebreaddev replied:
HttpResponse 객체로 해당 책임을 이전하였습니다.

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/HttpEntity.java:
```diff
@@ -0,0 +1,122 @@
+
+            final File file = new File(Objects.requireNonNull(resource).getFile());
+            final Path path = file.toPath();
+            final byte[] content = Files.readAllBytes(path);
```
반복되는 로직은 추상화를 해봐도 좋을 것 같네요!

jivebreaddev replied:
반복되는 로직도 추상화 하였습니다.

catsbi reviewed (changes requested):
시창님 4단계 리팩토링 미션 잘 구현해주셨어요. 늘 귀찮은데도 문서화 잘 해주시고, 여러 방법 고민하고 리팩토링 하시는 모습 인상깊네요 👍 

몇 가지 피드백 남겨드렸는데 확인 후 다시 리뷰요청 부탁드려요!

jivebreaddev committed:
feature(MapUtil): 피드백사항 - Stream 이용해서 OOM 회피, 반복되는 로직은 메서드 추출

jivebreaddev committed:
feature(MapUtil): 피드백사항 - Stream 이용해서 OOM 회피, 반복되는 로직은 메서드 추출
- IoUtils.copyLarge(inputStream,outputStream) 을 이용해서
Heap에 파일 저장하지 않고 진행하였으나, Content-length 의 값이
File.length() 와 같지 않음.

해당 문제는 HttpResponse 구성 요소마다 OutputStream을 나눠두고 content-length 를 측정해서 response header에 넣어야 하는것으로 보임.

jivebreaddev committed:
feature(MapUtil): 피드백사항 - Stream 이용해서 OOM 회피, 반복되는 로직은 메서드 추출
- IoUtils.copyLarge(inputStream,outputStream) 을 이용해서
Heap에 파일 저장하지 않고 진행하였으나, Content-length 의 값이
File.length() 와 같지 않음.

해당 문제는 HttpResponse 구성 요소마다 OutputStream을 나눠두고 content-length 를 측정해서 response header에 넣어야 하는것으로 보임.

jivebreaddev committed:
Style(RequestMapping): 피드백사항 - 인터페이스 이름과 구현체 같은 이름 제거
- RequestMapping 이라는 구현체 이름을 ControllerRequestMapping으로 변경하였습니다.

Jivebread committed:
Style(RequestMapping): 피드백사항 - set-cookie 일차함수로 리팩토링
- 명시적으로 Set-Cookie를 Cookie 에 추가하는 행동을 구현체로 생성하였습니다.

Jivebread committed:
feature(RequestMapping): 피드백사항 - ControllerConfig 가 n번 불렸을때, n개의 객체가 계속 재생성 되는 이슈

Jivebread committed:
feature(RequestMapping): 피드백사항 - HttpResponse의 빌더 사용해서 축약적 사용생성 + Response에서 HttpEntity 흡수
- StatusLine 축약 생성
- HttpEntity 로직 Response로 이동합니다.

Jivebread committed:
feature(RequestMapping): 피드백사항 - HttpResponse의 빌더 사용해서 축약적 사용생성 + Response에서 HttpEntity 흡수
- StatusLine 축약 생성
- HttpEntity 로직 Response로 이동합니다.

Jivebread committed:
feature(ControllerFactoryProviderConfig): 피드백사항 - factories 예외 컨트롤러?? 확장성 제한됨
- 예외 컨트롤러를 Factories 안으로 넣습니다.

Jivebread committed:
DOCS(README): 피드백사항 - Matcher의 matches, find 다른점 확인

jivebreaddev commented:
- [x] 추가 피드백 사항
  - [x] set-cookie 일차함수로 리팩토링 
      - https://github.com/next-step/java-http/pull/73#discussion_r1694030192
      - cookie 나 다른 cookie 를 설정할때, consumer 를 통해 설정하게끔 변경하였습니다.
  - [x] ControllerConfig 가 n번 불렸을때, n개의 객체가 계속 재생성 되는 이슈
  - [x] 인터페이스 이름과 구현체 같은 이름 제거
  - [x] Matcher의 matches, find 다른점 확인
    - matches는 전체의 String에서 regex 패턴을 매칭합니다.
    - 그와 별개로 find는 Substring 에서 Matching 을 확인합니다. -> /sdagagdex.dsifjas/index 
      - 이 경우에 find는 /index와 매칭이 됩니다.
  - [x] HttpResponse의 빌더 사용해서 축약적 사용생성
  - [x] factories 예외 컨트롤러?? 확장성 제한됨
  - [x] Response에서 HttpEntity 흡수
  - [x] 반복되는 로직은 메서드 추출 (HttpEntity)

  - [] readAllBytes 의 OOM 이슈로 제거하기 -> 여전히 HEAP에 파일을 읽어서 올리게 되는데 INPUTSTREAM -> OUTPUTSTREAM 으로 이전하는 과정에서 CONTENT-LENGTH 측정 실패 


톰캣에선 다음과 같이 Response에서 write 할때, content-length를 측정하는것같은데 output stream 을 write 할때,
@Functional Interface 로 contentLength를 업데이트 해줄려고했음. -> 아쉽지만 실패하였음


    public void doWrite(ByteBuffer chunk) throws IOException {
    int len = chunk.remaining();
    outputBuffer.doWrite(chunk);
    contentWritten += len - chunk.remaining();
    }

        public long getBytesWritten(boolean flush) {
        if (flush) {
            try {
                outputBuffer.flush();
            } catch (IOException ioe) {
                // Ignore - the client has probably closed the connection
            }
        }
        return getCoyoteResponse().getBytesWritten(flush);
    }

------------------------------------------------------------------------

한솔님의 피드백으로 더나은 코드를 작성하게되는것같아서 기분이 굉장히 좋습니다!

슬랙으로 실무 이야기도 해주시고 좋은 피드백들 너무나 감사합니다!

jivebreaddev committed:
Style(hTTP11Response): 피드백사항 - 인터페이스로 받는것 수정

catsbi reviewed (approved):
시창님, 피드백 고민 열심히 해주시고 반영 다 해주셔서 좋습니다 👍 
피드백 반영한 뒤 크게 조언드릴 부분들이 보이지 않아 머지하도록 하겠습니다. 바로 다음 5단계 진행 해주시죠!
